File: windows/ClaudeTrafficLight.Tests/UpdateCheckTests.cs
using ClaudeTrafficLight.Core;
using Xunit;

namespace ClaudeTrafficLight.Tests;

/// <summary>Mirror of the macOS UpdateCheckTests — the comparison rules must match.</summary>
public class UpdateCheckTests
{
    [Theory]
    [InlineData("1.1.0", new[] { 1, 1, 0 })]
    [InlineData("v1.2.3", new[] { 1, 2, 3 })]
    [InlineData(" V2.0 ", new[] { 2, 0 })]
    [InlineData("1.2.0-beta.1", new[] { 1, 2, 0 })]
    public void Parse_handles_plain_and_prefixed_versions(string input, int[] expected)
        => Assert.Equal(expected, UpdateCheck.Parse(input));

    [Theory]
    [InlineData("")]
    [InlineData("latest")]
    [InlineData("1.x")]
    public void Parse_rejects_non_numeric_input(string input)
        => Assert.Null(UpdateCheck.Parse(input));

    [Theory]
    [InlineData("v1.1.0", "1.0.1", true)]
    [InlineData("1.10.0", "1.9.9", true)]  // numeric, not lexicographic
    [InlineData("2.0", "1.99.99", true)]
    [InlineData("1.1.0", "1.1.0", false)]
    [InlineData("v1.1", "1.1.0", false)]   // padded equal
    [InlineData("1.0.9", "1.1.0", false)]
    [InlineData("banana", "1.0.0", false)] // malformed tag never nags
    [InlineData("v2.0.0", "", false)]      // dev build without a version never nags
    public void IsNewer_compares_numerically(string latest, string current, bool expected)
        => Assert.Equal(expected, UpdateCheck.IsNewer(latest, current));

    [Fact]
    public void ParseLatestRelease_extracts_tag_and_url()
    {
        var parsed = UpdateCheck.ParseLatestRelease(
            "{\"tag_name\":\"v1.1.0\",\"html_url\":\"https://github.com/merve/claude-traffic-light/releases/tag/v1.1.0\"}");
        Assert.Equal("v1.1.0", parsed?.Tag);
        Assert.Equal("https://github.com/merve/claude-traffic-light/releases/tag/v1.1.0", parsed?.Url);
    }

    [Fact]
    public void ParseLatestRelease_falls_back_to_releases_page_without_html_url()
        => Assert.Equal(UpdateCheck.ReleasesPage,
                        UpdateCheck.ParseLatestRelease("{\"tag_name\":\"v9.9.9\"}")?.Url);

    [Theory]
    [InlineData("not json")]
    [InlineData("{}")]
    public void Garbage_payload_returns_null(string json)
        => Assert.Null(UpdateCheck.ParseLatestRelease(json));
}

File: macos/Tests/ClaudeStatusCoreTests/UpdateCheckTests.swift
import XCTest
@testable import ClaudeStatusCore

final class UpdateCheckTests: XCTestCase {

    func testParseHandlesPlainAndPrefixedVersions() {
        XCTAssertEqual(UpdateCheck.parse("1.1.0"), [1, 1, 0])
        XCTAssertEqual(UpdateCheck.parse("v1.2.3"), [1, 2, 3])
        XCTAssertEqual(UpdateCheck.parse(" V2.0 "), [2, 0])
        XCTAssertEqual(UpdateCheck.parse("1.2.0-beta.1"), [1, 2, 0])
        XCTAssertNil(UpdateCheck.parse(""))
        XCTAssertNil(UpdateCheck.parse("latest"))
        XCTAssertNil(UpdateCheck.parse("1.x"))
    }

    func testIsNewerComparesNumerically() {
        XCTAssertTrue(UpdateCheck.isNewer(latest: "v1.1.0", current: "1.0.1"))
        XCTAssertTrue(UpdateCheck.isNewer(latest: "1.10.0", current: "1.9.9")) // not lexicographic
        XCTAssertTrue(UpdateCheck.isNewer(latest: "2.0", current: "1.99.99"))
        XCTAssertFalse(UpdateCheck.isNewer(latest: "1.1.0", current: "1.1.0"))
        XCTAssertFalse(UpdateCheck.isNewer(latest: "v1.1", current: "1.1.0")) // padded equal
        XCTAssertFalse(UpdateCheck.isNewer(latest: "1.0.9", current: "1.1.0"))
    }

    // A malformed tag or a dev build without a version must never nag about updates.
    func testUnparseableInputIsNeverNewer() {
        XCTAssertFalse(UpdateCheck.isNewer(latest: "banana", current: "1.0.0"))
        XCTAssertFalse(UpdateCheck.isNewer(latest: "v2.0.0", current: ""))
    }

    func testParseLatestReleasePayload() {
        let json = #"{"tag_name":"v1.1.0","html_url":"https://github.com/merve/claude-traffic-light/releases/tag/v1.1.0"}"#
        let parsed = UpdateCheck.parseLatestRelease(json: Data(json.utf8))
        XCTAssertEqual(parsed?.tag, "v1.1.0")
        XCTAssertEqual(parsed?.url.absoluteString,
                       "https://github.com/merve/claude-traffic-light/releases/tag/v1.1.0")
    }

    func testParseLatestReleaseFallsBackToReleasesPageWithoutHtmlUrl() {
        let parsed = UpdateCheck.parseLatestRelease(json: Data(#"{"tag_name":"v9.9.9"}"#.utf8))
        XCTAssertEqual(parsed?.url, UpdateCheck.releasesPage)
    }

    func testGarbagePayloadReturnsNil() {
        XCTAssertNil(UpdateCheck.parseLatestRelease(json: Data("not json".utf8)))
        XCTAssertNil(UpdateCheck.parseLatestRelease(json: Data("{}".utf8)))
    }
}
